apm-agent-plugins/apm-jms-plugin/apm-jms-plugin-base/src/main/java/co/elastic/apm/agent/jms/JmsInstrumentationHelper.java

```diff
@@ -143,16 +143,16 @@ public Span startJmsSendSpan(Destination destination, Message message) {
     }
 
     @Nullable
-    public Transaction startJmsTransaction(Message parentMessage, Class<?> instrumentedClass) {
-        Transaction transaction = tracer.startChildTransaction(parentMessage, JmsMessagePropertyAccessor.instance(), PrivilegedActionUtils.getClassLoader(instrumentedClass));
+    public Transaction<?> startJmsTransaction(Message parentMessage, Class<?> instrumentedClass) {
+        Transaction<?> transaction = tracer.startChildTransaction(parentMessage, JmsMessagePropertyAccessor.instance(), PrivilegedActionUtils.getClassLoader(instrumentedClass));
         if (transaction != null) {
             transaction.setFrameworkName(FRAMEWORK_NAME);
         }
         return transaction;
     }
 
-    public void makeChildOf(Transaction childTransaction, Message parentMessage) {
-        TraceContext.<Message>getFromTraceContextTextHeaders().asChildOf(childTransaction.getTraceContext(), parentMessage, JmsMessagePropertyAccessor.instance());
+    public void makeChildOf(Transaction<?> childTransaction, Message parentMessage) {
+        childTransaction.addLink(JmsMessagePropertyAccessor.instance(), parentMessage);
```

Comment:
I can confirm that this is the only such use within the agent project. That is why I was wondering about it while not feeling sufficiently competent to implement the change. Thanks for any feedback on this!

Reply:
Yes, this is a unique scenario that definitely complicates stuff 😐 
And for such
It is related to the tricky nature of monitoring polling APIs for messaging systems (like `receive`)- the only thing that is interesting in these operations is what happens _after they end_ - the handling of the message they return. When instrumenting `onMessage` for example, it's easy because the instrumented method encapsulates what we are interested at. In other messaging systems we found a descent way of handling poling APIs (meaning - tracing stuff outside instrumented methods), but there is no such in JMS. So what we do is a much more dangerous thing than @JonasKunz pointed out (😬 ) - we start a transaction that encapsulates the massage handling after `receive` ends and ending the transaction when the next `receive` starts, assuming that messages are read in dedicated threads and message handling is either done on the same thread, or our cross-thread context propagation will take care of that. This is why `receiveNoWait` gets a different handling- it doesn't fit a pattern of message-reading thread.
Since the capturing of "message-handling" transactions is based on this "dangerous" heuristic, we added a way to shut it off and capture transactions of only the message receive, to at least get a minimal distributed trace and help us and users analyze or improve later. This is controlled through the [`message_polling_transaction_strategy `](https://github.com/elastic/apm-agent-java/blob/7d328022a4b3fe3e7ede19bfad1cd69f9bf738a1/apm-agent-core/src/main/java/co/elastic/apm/agent/configuration/MessagingConfiguration.java#L36)

Bottom line:
1. these "polling transactions" you see here are off by default and I am not familiar with a place where they were needed (as far as I know our message-reading-thread heuristic worked so far, including in the Spring implementation)
2. there is no real risk of child spans, as the instrumented methods normally only block and return, but not activating makes sense
3. span link is preferable in almost any way, except from the fact that in our UI this currently won't show up in service map and dependencies table. Since this is such a corner case, I'd say- let's replace it with a span link, but please rename this helper method to something else, even `addSpanLink`

Reply:
I removed the helper method as I found it to be better readable if it is placed in the code directly.